File: 02-python3_core/04-decorators/demo_timer.cpp
#include <chrono>
#include <iostream>
#include <thread>

// Define the decorator equivalent
template <typename Func>
auto execution_time(Func func) {
    return [func]() {
        // Record the start time
        auto start_time = std::chrono::high_resolution_clock::now();

        // Call the original function
        auto result = func();

        // Record the end time
        auto end_time = std::chrono::high_resolution_clock::now();

        // Calculate execution time
        auto duration =
            std::chrono::duration<double>(end_time - start_time).count();
        std::cout << "Function executed in " << duration << " seconds"
                  << std::endl;

        return result;
    };
}

// Define a function to simulate slow operation
std::string slow_function() {
    std::this_thread::sleep_for(
        std::chrono::seconds(2));  // Simulate a slow operation
    return "Finished";
}

int main() {
    // Use the decorator equivalent
    auto decorated_function = execution_time(slow_function);

    // Call the decorated function
    std::cout << decorated_function() << std::endl;

    return 0;
}

File: 04-python3_core/04-decorators/demo_timer.py
import time

# Define the decorator
def execution_time(func):
    def wrapper(*args, **kwargs):
        start_time = time.time()  # Record the start time
        result = func(*args, **kwargs)  # Call the original function
        end_time = time.time()  # Record the end time
        print(f"Function '{func.__name__}' executed in {end_time - start_time:.4f} seconds")
        return result
    return wrapper

# Use the decorator
@execution_time
def slow_function():
    time.sleep(2)  # Simulate a slow operation
    return "Finished"

# Call the decorated function
print(slow_function())
